File: src/main/java/com/tim/sernachat/config/StompHandler.java
package com.tim.sernachat.config;

import com.tim.sernachat.model.ChatMessage;
import com.tim.sernachat.repository.ChatRoomRepository;
import com.tim.sernachat.service.ChatService;
import com.tim.sernachat.service.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.messaging.Message;
import org.springframework.messaging.MessageChannel;
import org.springframework.messaging.simp.stomp.StompCommand;
import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
import org.springframework.messaging.support.ChannelInterceptor;
import org.springframework.stereotype.Component;

import java.security.Principal;
import java.util.Optional;

@Slf4j
@Component
@RequiredArgsConstructor
public class StompHandler implements ChannelInterceptor {

    private final JwtTokenProvider jwtTokenProvider;
    private final ChatRoomRepository chatRoomRepository;
    private final ChatService chatService;

    @Override
    public Message<?> preSend(Message<?> message, MessageChannel channel) {
        StompHeaderAccessor accessor = StompHeaderAccessor.wrap(message);
        if (StompCommand.CONNECT == accessor.getCommand()) {
            String jwtToken = accessor.getFirstNativeHeader("token");
            log.info("CONNECT {}", jwtToken);
            jwtTokenProvider.validateToken(jwtToken);
        } else if (StompCommand.SUBSCRIBE == accessor.getCommand()) {
            String roomId = chatService.getRoomId(Optional.ofNullable((String) message.getHeaders().get("simpDestination")).orElse("InvalidRoomId"));
            String sessionId = (String) message.getHeaders().get("simpSessionId");
            chatRoomRepository.setUserEnterInfo(sessionId, roomId);
            chatRoomRepository.plusUserCount(roomId);
            String name = Optional.ofNullable((Principal) message.getHeaders().get("simpUser")).map(Principal::getName).orElse("UnknownUser");
            chatService.sendChatMessage(ChatMessage.builder().type(ChatMessage.MessageType.ENTER).roomId(roomId).from(name).build());
            log.info("SUBSCRIBED {}, {}", name, roomId);

        } else if (StompCommand.DISCONNECT == accessor.getCommand()) {
            String sessionId = (String) message.getHeaders().get("simpSessionId");
            String roomId = chatRoomRepository.getUserEnterRoomId(sessionId);

            chatRoomRepository.minusUserCount(roomId);

            String name = Optional.ofNullable((Principal) message.getHeaders().get("simpUser")).map(Principal::getName).orElse("UnknownUser");
            chatService.sendChatMessage(ChatMessage.builder().type(ChatMessage.MessageType.QUIT).roomId(roomId).from(name).build());

            chatRoomRepository.removeUserEnterInfo(sessionId);
            log.info("DISCONNECTED {}, {}", sessionId, roomId);
        }
        return message;
    }
}

File: src/main/kotlin/com/itime/sernachat/service/StompHandler.kt
package com.itime.sernachat.service

//import com.itime.sernachat.repository.1]ChatRoomRepository
import org.slf4j.Logger
import org.springframework.messaging.Message
import org.springframework.messaging.MessageChannel
import org.springframework.messaging.support.ChannelInterceptor
import org.springframework.stereotype.Component

@Component
class StompHandler(
    val logger: Logger,
    val jwtTokenProvider: JwtTokenProvider,
//    val chatRoomRepository: ChatRoomRepository,
    val chatService: ChatService
) : ChannelInterceptor {

    override fun preSend(message: Message<*>, channel: MessageChannel): Message<*> {
//        val accessor = StompHeaderAccessor.wrap(message)
//        if (StompCommand.CONNECT == accessor.command) {

//            val jwtToken = accessor.getFirstNativeHeader("token")!!
//            logger.info("CONNECT {}", jwtToken)
//            jwtTokenProvider.validateToken(jwtToken)

//        } else if (StompCommand.SUBSCRIBE == accessor.command) {

//            val roomId = "123"
//            val sessionId = message.headers["simpSessionId"] as String
//            chatRoomRepository.setUserEnterInfo(sessionId, roomId)
//            chatRoomRepository.plusUserCount(roomId)
//            val name = Optional.ofNullable(message.headers["simpUser"] as Principal?).map { obj: Principal -> obj.name }
//                .orElse("UnknownUser")
//            chatService.sendChatMessage(
//                CommonMessage.type(CommonMessage.MessageType.ENTER).chatId(roomId).sender(name).build()
//            )
//            logger.info("SUBSCRIBED {}, {}", name, roomId)
//        } else if (StompCommand.DISCONNECT == accessor.command) {

//            val sessionId = message.headers["simpSessionId"] as String
//            val roomId = chatRoomRepository.getUserEnterRoomId(sessionId)!!
//            chatRoomRepository.minusUserCount(roomId)
//            val name = Optional.ofNullable(message.headers["simpUser"] as Principal?).map { obj: Principal -> obj.name }
//                .orElse("UnknownUser")
//            chatService.sendChatMessage(
//                CommonMessage.builder().type(CommonMessage.MessageType.QUIT).roomId(roomId).from(name).build()
//            )
//            chatRoomRepository.removeUserEnterInfo(sessionId)
//            logger.info("DISCONNECTED {}, {}", sessionId, roomId)
//        }
        return message
    }
}